### PR TITLE
Proxy support

### PR DESCRIPTION
**Pull request by bmanojlovic**
add  proxy support (https://www.haproxy.com/blog/haproxy/proxy-protocol/)
(currently proxylistener is my forked code but i hope it will be accepted in upstream)

### CODE REVIEW

**Comment by bmanojlovic** [comment]
will cleanup history so please ignore level of commits (squash)
please comment if you find it acceptable in current state or wanted something to extend a bit

**Comment by jhillyerd** [comment]
Thanks.  I don't use HAProxy myself -- if I understand correctly the primary benefit of this change is to allow Inbucket to act-on and/or log the actual client source IP, instead of that of the load balancers IP?

I am going to retarget this at the `develop` branch (we are still using git-flow until 3.0.0 release, see CONTRIBUTING.md.)
Given I am very close to releasing 3.0.0, this particular feature will have to go in 3.1.0.  We can either let it hang out in this PR, or create a special feature branch that I can keep rebased until 3.0.0 is out the door.  Let me know if you have a preference on that.  BTW, GitHub can squash PRs when merging, and I almost always use that feature, so you don't need to worry about squashing.

1. Do you know if we have any options for unit testing this without standing up an HAProxy server?  There is no QA department, so I rely heavily on our tests to know if I've broken something. :)

2. We will need to update `doc/config.md` for the new option.

File: etc/dev-start.sh [hunk 1]
```diff
@@ -19,6 +19,9 @@ export INBUCKET_STORAGE_PARAMS="path:/tmp/inbucket"
 export INBUCKET_STORAGE_RETENTIONPERIOD="3h"
 export INBUCKET_STORAGE_MAILBOXMSGCAP="300"
 
+
+export INBUCKET_SMTP_PROXYHOSTS="127.0.0.1"
```
Comment (jhillyerd) [inline etc/dev-start.sh:23]: Let's move this up alphabetically with the other SMTP options, and leave commented out by default.

File: pkg/config/config.go [hunk 2]
```diff
@@ -70,6 +70,7 @@ type Root struct {
 type SMTP struct {
 	Addr            string        `required:"true" default:"0.0.0.0:2500" desc:"SMTP server IP4 host:port"`
 	Domain          string        `required:"true" default:"inbucket" desc:"HELO domain"`
+	ProxyHosts      []string      `desc:"Accept network connection from PROXY host(s)"`
```
Comment (jhillyerd) [inline pkg/config/config.go:73]: I'm not sure how proxylistener works, would it be more crrect to say "Only accept network connections from PROXY host(s)" ?

File: pkg/server/smtp/listener.go [hunk 3]
```diff
@@ -9,6 +9,7 @@ import (
 	"sync"
 	"time"
 
+	proxylistener "github.com/anisse/proxylistener"
```
Comment (jhillyerd) [inline pkg/server/smtp/listener.go:12]: The import alias looks redundant to me.

File: pkg/server/smtp/listener.go [hunk 4]
```diff
@@ -133,6 +138,8 @@ func (s *Server) serve(ctx context.Context) {
 	var tempDelay time.Duration
 	for sessionID := 1; ; sessionID++ {
 		if conn, err := s.listener.Accept(); err != nil {
+			slog := log.With().Str("proxy", "false").Logger()
```
Comment (jhillyerd) [inline pkg/server/smtp/listener.go:141]: Hard coded false does not look useful?

File: pkg/server/smtp/listener.go [hunk 5]
```diff
@@ -133,6 +138,8 @@ func (s *Server) serve(ctx context.Context) {
...
+			slog.Debug().Msg("s.listener.")
```
Comment (jhillyerd) [inline pkg/server/smtp/listener.go:142]: This also does not look useful.

**Review by jhillyerd — CHANGES_REQUESTED** [review]

File: pkg/server/smtp/listener.go (hunk 3 above)
Comment (bmanojlovic) [reply pkg/server/smtp/listener.go:12]: agreed

File: pkg/config/config.go (hunk 2 above)
Comment (bmanojlovic) [reply pkg/config/config.go:73]: That is correct, as any other host connecting and using "proxy protocol" could fake connection connecting ip.

File: etc/dev-start.sh (hunk 1 above)
Comment (bmanojlovic) [reply etc/dev-start.sh:23]: agreed

File: pkg/server/smtp/listener.go (hunk 4 above)
Comment (bmanojlovic) [reply pkg/server/smtp/listener.go:141]: yes that can go away - it was really code used for debugging

File: pkg/server/smtp/listener.go (hunk 5 above)
Comment (bmanojlovic) [reply pkg/server/smtp/listener.go:142]: same

**Comment by bmanojlovic** [comment]
I will need to think what would be easiest way to test it ( cleaned up commit history for now in any case )

File: pkg/config/config.go (hunk 2 above)
Comment (bmanojlovic) [reply pkg/config/config.go:73]: hm i need to fix that part... it is broken :)

Comment (bmanojlovic) [reply pkg/config/config.go:73]:
```
6:46PM INF Starting SMTP session module=smtp remote=10.10.10.10:2222 session=1
```
from different host ... i need to check who is remote host (this was checked from remote host that should not be able to use proxy protocol...)
and this is by the way way to test if proxy protocol works (log entry)

**Comment by jhillyerd** [comment]
Don't worry about the tests if you can't think of a good way to do it.  I would like to add an integration test suite to Inbucket to test SMTP, POP3 and REST APIs.  PROXY support could be part of that eventually.

Once you feel the rest of this PR is working correctly, toggle it off Draft and I'll take another look.

**Pull request closed without merging**